File: airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/util/JavaProcessRunner.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.s3.util;

import io.airbyte.commons.io.LineGobbler;
import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.concurrent.TimeUnit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class JavaProcessRunner {

  private static final Logger LOGGER = LoggerFactory.getLogger(JavaProcessRunner.class);

  public static void runProcess(final String path, final Runtime run, final String... commands) throws IOException, InterruptedException {
    LOGGER.info("Running process: " + Arrays.asList(commands));
    final Process pr = path.equals(System.getProperty("user.dir")) ? run.exec(commands) : run.exec(commands, null, new File(path));
    LineGobbler.gobble(pr.getErrorStream(), LOGGER::error);
    LineGobbler.gobble(pr.getInputStream(), LOGGER::info);
    if (!pr.waitFor(10, TimeUnit.MINUTES)) {
      pr.destroy();
      throw new RuntimeException("Timeout while executing: " + Arrays.toString(commands));
    }
  }

}

File: airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/util/JavaProcessRunner.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.destination.s3.util

import io.airbyte.commons.io.LineGobbler
import io.github.oshai.kotlinlogging.KotlinLogging
import java.io.File
import java.io.IOException
import java.util.*
import java.util.concurrent.TimeUnit

private val LOGGER = KotlinLogging.logger {}

object JavaProcessRunner {

    @Throws(IOException::class, InterruptedException::class)
    fun runProcess(path: String, run: Runtime, vararg commands: String?) {
        LOGGER.info { "Running process: ${listOf(*commands)}" }
        val pr =
            if (path == System.getProperty("user.dir")) run.exec(commands)
            else run.exec(commands, null, File(path))
        LineGobbler.gobble(`is` = pr.errorStream, { LOGGER.warn { it } })
        LineGobbler.gobble(`is` = pr.inputStream, { LOGGER.info { it } })
        if (!pr.waitFor(10, TimeUnit.MINUTES)) {
            pr.destroy()
            throw RuntimeException("Timeout while executing: " + commands.contentToString())
        }
    }
}
